lib/elixir/lib/gen_server.ex

```diff
@@ -566,6 +566,10 @@ defmodule GenServer do
   """
   @spec call(server, term, timeout) :: term
   def call(server, request, timeout \\ 5000) do
+    if server == self do
+      raise ArgumentError, message: 
+      "invalid server argument in call(server, request, timeout). Looks like you trying to call a parent process"
```

Comment:
We should probably use `exit {:calling_self, {:gen_server, :call, [server, request, timeout]}` instead of raise to not change the failure semantics and use our exception layer to translate it to something meaningful. /cc @fishcakez